crates/cxx-qt-gen/src/lib.rs

```diff
@@ -73,45 +78,84 @@ mod tests {
         code
     }
 
-    /// Helper for testing if a given input Rust file generates the expected C++ & Rust code
-    /// This needs to be a macro rather than a function because include_str needs the file path at compile time.
-    macro_rules! test_code_generation {
-        ( $file_stem:literal ) => {
-            let parser = Parser::from(
-                syn::parse_str(include_str!(concat!("../test_inputs/", $file_stem, ".rs")))
-                    .unwrap(),
-            )
-            .unwrap();
+    fn update_expected_file(path: PathBuf, source: &str) {
+        println!("Updating expected file: {:?}", path);
 
-            let generated_cpp = GeneratedCppBlocks::from(&parser).unwrap();
-            let (header, source) =
-                if let CppFragment::Pair { header, source } = write_cpp(&generated_cpp) {
-                    (sanitize_code(header), sanitize_code(source))
-                } else {
-                    panic!("Expected CppFragment::Pair")
-                };
-            let expected_cpp_header =
-                clang_format(include_str!(concat!("../test_outputs/", $file_stem, ".h")))
-                    .map(sanitize_code)
-                    .unwrap();
-            let expected_cpp_source = clang_format(include_str!(concat!(
-                "../test_outputs/",
-                $file_stem,
-                ".cpp"
-            )))
-            .map(sanitize_code)
+        let mut file = OpenOptions::new()
+            .write(true)
+            .truncate(true)
+            .open(path)
             .unwrap();
+        file.write_all(source.as_bytes()).unwrap();
+    }
+
+    fn update_expected(test_name: &str, rust: &str, header: &str, source: &str) -> bool {
+        // Ideally we'd be able to get the path from `file!()`, but that unfortunately only
+        // gives us a relative path, which isn't really all that useful.
+        // So require the path of the crate to be set via an environment variable.
+        //
+        // In the simplest case this can be achieved by running:
+        //
+        //      CXXQT_UPDATE_EXPECTED=$(pwd) cargo test
```

Comment:
Probably is the simplest way of doing it as this is a developer tool.

Assume that making a bin target or having a separate crate that we don't publish which updates the files would be way over the top and could leak into crates.io as a installable thing via cargo. And would add more complexity.

Reply:
Yeah, I'm fine with the Environment variable approch per se.
Easy to implement & maintain and will continue to work as we add new tests.
I was hoping to not have to provide the path to it though.

I.e. instead of having to go:
```
CXXQT_UPDATE_EXPECTED=/path/to/crates/cxx-qt-gen
```
I would have liked:
```
CXXQT_UPDATE_EXPECTED=1
```

We could just add a small shell script that does the updating though.
``` bash
#!/bin/bash

set -ex

SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
cd ${SCRIPT_DIR}
CXXQT_UPDATE_EXPECTED="${SCRIPT_DIR}" cargo test tests::generates
```

Reply:
Done